File: WebMacroSoftKeyboard/ClientAppFlutter/web_macro_soft_keyboard_client/lib/models/ConfigParameter.dart
import 'dart:ffi';

// class ConfigParameter {
//   final int id;
//   final ConfigParameterType configParameterType;
//   final String symbolicName;
//   final String value;

//   ConfigParameter({
//     required this.id,
//     required this.configParameterType,
//     required this.symbolicName,
//     required this.value,
//   });

//   ConfigParameter.fromJson(Map<String, dynamic> json)
//       : id = json["id"],
//         configParameterType = json["configParameterType"],
//         symbolicName = json["symbolicName"],
//         value = json["pluginNvalueame"];
// }

// enum ConfigParameterType { Control, View }

class ConfigParameter {
  final ConfigParameterType configParameterType;
  final String symbolicName;
  final Object defaultValue;
  final String? validationRegEx;
  final int? minValue;
  final int? maxValue;

  ConfigParameter({
    required this.configParameterType,
    required this.symbolicName,
    required this.defaultValue,
    required this.validationRegEx,
    required this.minValue,
    required this.maxValue,
  });

  ConfigParameter.fromJson(Map<String, dynamic> json)
      : configParameterType = ConfigParameterType.from[json["parameterType"]],
        symbolicName = json["symbolicName"],
        defaultValue = json["defaultValue"],
        validationRegEx = json["validationRegEx"],
        minValue = json["minValue"],
        maxValue = json["maxValue"];
}

enum ConfigParameterType {
  from,
  string,
  int,
  bool,
}

extension ConfigParameterTypeIndex on ConfigParameterType {
  operator [](String key) => (name) {
        switch (name) {
          case 'string':
            return ConfigParameterType.string;
          case 'int':
            return ConfigParameterType.int;
          case 'bool':
            return ConfigParameterType.bool;
          default:
            throw RangeError("enum ConfigParameterType contains no value '$name'");
        }
      }(key);
}

File: MakroBoard/ApiModels/ConfigParameter.cs
namespace MakroBoard.ApiModels
{
    public class ConfigParameter
    {
        private ConfigParameter(string symbolicName, LocalizableString label)
        {
            SymbolicName = symbolicName;
            Label = label;
        }

        public ConfigParameter(string symbolicName, LocalizableString label, string defaultValue, string validationRegEx) : this(symbolicName, label)
        {
            ParameterType = "string";
            ValidationRegEx = validationRegEx;
            DefaultValue = defaultValue;
        }

        public ConfigParameter(string symbolicName, LocalizableString label, int minValue, int maxValue) : this(symbolicName, label)
        {
            ParameterType = "int";
            MinValue = minValue;
            MaxValue = maxValue;
        }

        public ConfigParameter(string symbolicName, LocalizableString label, bool defaultValue) : this(symbolicName, label)
        {
            ParameterType = "bool";
            DefaultValue = defaultValue;
        }

        public string SymbolicName { get; }
        public LocalizableString Label { get; }
        public string ParameterType { get; }
        public object DefaultValue { get; }
        public string ValidationRegEx { get; }
        public int MinValue { get; }
        public int MaxValue { get; }
    }
}
